bdk_core/src/io.rs

```diff
@@ -0,0 +1,249 @@
+                let mut buf = [0u8; 36];
+                u32::to_be_bytes(*height).iter().enumerate().for_each(|(i, val)| buf[i] = *val);
+                encode::serialize(block_hash).iter().enumerate().for_each(|(i, val)| buf[i + 4] = *val);
+                let _size = cursor.write(&buf)?;
+            } else {
+                let zero_pad_buf = [0u8; 36];
+                let _size = cursor.write(&zero_pad_buf)?;
+            }
+        }
+    }
+
+    for (height, txid) in &chain.txid_by_height {
+        let mut buf = match height {
+            TxHeight::Confirmed(block_height)=> u32::to_be_bytes(*block_height).to_vec(),
+            TxHeight::Unconfirmed => u32::to_be_bytes(UNCONFIRMED_TX_HEIGHT).to_vec()
+        };
+        buf.extend(encode::serialize(txid));
+        let _size = cursor.write(&buf)?;
+    }
+
+    Ok(())
+}
+
+/// Instantiate SparseChain from disk bytes.
+pub fn instantiate_sparse_chain<T: Read + Seek>(chain: &mut SparseChain, reader: &mut T) -> Result<(), IOError> {
+    let bytes = reader.bytes().collect::<io::Result<Vec<u8>>>()?;
+    let mut cursor = io::Cursor::new(bytes);
+    // skip the state_id + version
+    cursor.set_position(8);
+    let mut checkpoint_limit_bytes = [0u8; 4];
+    let _size = cursor.read(&mut checkpoint_limit_bytes)?;
+    let checkpoint_limit = u32::from_be_bytes(checkpoint_limit_bytes);
+    let mut checkpoint_bytes = Vec::with_capacity((checkpoint_limit as u32 * 36u32) as usize);
+    // read checkpoint_limit * 36 bytes
+    let _size = cursor.read(&mut checkpoint_bytes)?;
+    // create checkpoints from these bytes.
+    for chunk in checkpoint_bytes.chunks(36) {
+        // once you meet the padded zeros stop reading
+        if chunk.iter().all(|x| *x == 0u8) {
+            break;
+        }
+
+        let height = u32::from_be_bytes(<[u8; 4]>::try_from(&chunk[0..4])?);
+        let block_hash: BlockHash = encode::deserialize(&chunk[4..])?;
+        chain.checkpoints.insert(height, block_hash);
+    }
+
+    // read (height + txid) data
+    let mut txid_bytes = Vec::new();
+    // create (height, txid) mapping and the reverse mapping
+    let _size = cursor.read_to_end(&mut txid_bytes)?;
+    for chunk in txid_bytes.chunks(36) {
+        let height = u32::from_be_bytes(<[u8; 4]>::try_from(&chunk[0..4])?);
+        let txid: Txid = encode::deserialize(&chunk[4..])?;
+        let _set = if height == (UNCONFIRMED_TX_HEIGHT) {
+            chain.insert_tx(txid.clone(), TxHeight::Unconfirmed)?
+        } else {
+            chain.insert_tx(txid.clone(), TxHeight::Confirmed(height))?
+        };
+    }
+    Ok(())
+}
+
+/// write TxGraph to disk for persistence storage
+pub fn write_tx_graph<T: Write + Seek>(
+    tx_graph: &TxGraph,
+    writer: &mut T
+) -> io::Result<()> {
+    // TODO: write current state_id + version
+    let header_bytes = [0u8; 8];
+    let _size = writer.write(&header_bytes)?;
+    let mut buf = Vec::new();
```

Comment:
I first wrote to a vec because I imagine it will be expensive to make so many system calls rather than just writing to memory and making one system call for writing at the end. Now that I think about it, maybe using a buffered writer here will make much sense.